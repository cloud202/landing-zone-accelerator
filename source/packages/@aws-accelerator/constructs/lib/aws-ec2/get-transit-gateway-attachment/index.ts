/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as AWS from 'aws-sdk';

import { throttlingBackOff } from '@aws-accelerator/utils';

AWS.config.logger = console;

/**
 * get-transit-gateway-attachment - lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string | undefined;
      Status: string;
    }
  | undefined
> {
  const name = event.ResourceProperties['name'];
  const transitGatewayId = event.ResourceProperties['transitGatewayId'];
  const roleArn = event.ResourceProperties['roleArn'];

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      const stsClient = new AWS.STS({});

      const assumeRoleResponse = await throttlingBackOff(() =>
        stsClient
          .assumeRole({
            RoleArn: roleArn,
            RoleSessionName: 'GetTransitGatewayAttachmentSession',
          })
          .promise(),
      );

      const ec2Client = new AWS.EC2({
        credentials: {
          accessKeyId: assumeRoleResponse.Credentials?.AccessKeyId ?? '',
          secretAccessKey: assumeRoleResponse.Credentials?.SecretAccessKey ?? '',
          sessionToken: assumeRoleResponse.Credentials?.SessionToken,
        },
      });

      let nextToken: string | undefined = undefined;
      do {
        const page = await throttlingBackOff(() =>
          ec2Client.describeTransitGatewayAttachments({ NextToken: nextToken }).promise(),
        );
        for (const attachment of page.TransitGatewayAttachments ?? []) {
          if (attachment.TransitGatewayId === transitGatewayId && attachment.State === 'available') {
            const nameTag = attachment.Tags?.find(t => t.Key === 'Name');
            if (nameTag && nameTag.Value === name) {
              console.log(attachment);
              return {
                PhysicalResourceId: attachment.TransitGatewayAttachmentId,
                Status: 'SUCCESS',
              };
            }
          }
        }
        nextToken = page.NextToken;
      } while (nextToken);

      throw new Error(`Attachment ${name} for ${transitGatewayId} not found`);

    case 'Delete':
      // Do Nothing
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}
