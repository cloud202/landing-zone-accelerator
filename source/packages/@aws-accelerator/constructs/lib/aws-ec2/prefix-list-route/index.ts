/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

/**
 * aws-ec2-prefix-list-route - lambda handler
 *
 * @param event
 * @returns
 */

import * as AWS from 'aws-sdk';

import { throttlingBackOff } from '@aws-accelerator/utils';

const ec2 = new AWS.EC2();
export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent): Promise<
  | {
      Status: string | undefined;
    }
  | undefined
> {
  interface RouteProps {
    readonly DestinationPrefixListId: string;
    readonly RouteTableId: string;
    readonly CarrierGatewayId?: string;
    readonly EgressOnlyInternetGatewayId?: string;
    readonly GatewayId?: string;
    readonly InstanceId?: string;
    readonly LocalGatewayId?: string;
    readonly NatGatewayId?: string;
    readonly NetworkInterfaceId?: string;
    readonly TransitGatewayId?: string;
    readonly VpcEndpointId?: string;
    readonly VpcPeeringConnectionId?: string;
  }

  const props: RouteProps = event.ResourceProperties['routeDefinition'];

  switch (event.RequestType) {
    case 'Create':
      await throttlingBackOff(() => ec2.createRoute(props).promise());

      return {
        Status: 'SUCCESS',
      };

    case 'Update':
      await throttlingBackOff(() => ec2.replaceRoute(props).promise());

      return {
        Status: 'SUCCESS',
      };

    case 'Delete':
      await throttlingBackOff(() =>
        ec2
          .deleteRoute({ DestinationPrefixListId: props.DestinationPrefixListId, RouteTableId: props.RouteTableId })
          .promise(),
      );

      return {
        Status: 'SUCCESS',
      };
  }
}
