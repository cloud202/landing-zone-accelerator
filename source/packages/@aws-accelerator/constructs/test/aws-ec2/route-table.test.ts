/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { RouteTable } from '../../lib/aws-ec2/route-table';
import { Vpc } from '../../lib/aws-ec2/vpc';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(RouteTable): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

const vpc = new Vpc(stack, 'TestVpc', {
  name: 'Test',
  ipv4CidrBlock: '10.0.0.0/16',
  internetGateway: true,
  enableDnsHostnames: false,
  enableDnsSupport: true,
  instanceTenancy: 'default',
});

new RouteTable(stack, 'RouteTable', {
  name: 'TestRouteTable',
  vpc: vpc,
  tags: [{ key: 'Test-Key', value: 'Test-Value' }],
});

/**
 * RouteTable construct test
 */
describe('RouteTable', () => {
  snapShotTest(testNamePrefix, stack);
});
