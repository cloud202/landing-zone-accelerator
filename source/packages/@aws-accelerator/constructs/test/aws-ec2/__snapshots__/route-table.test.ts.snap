// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteTable Construct(RouteTable):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "RouteTable82FB8FA6": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestRouteTable",
          },
          Object {
            "Key": "Test-Key",
            "Value": "Test-Value",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpcE77CE678": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": false,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "TestVpcInternetGateway01360C82": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "TestVpcInternetGatewayAttachment60E451D5": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "TestVpcInternetGateway01360C82",
        },
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
}
`;
